docs: Add compound component doc (#937)
Workday/canvas-kit: COMPOUND_COMPONENTS.md

## Models

### What is a Model?

 If a compound component was stripped of all its markup, attributes, and styling, what would remain is the model. The model is how we describe the state and behavior of the component. You could completely swap out the underlying elements, attributes, and styles, and the model would remain the same. The model is an object that is composed of two parts: `state` and `events`. The model's `state` describes the current snapshot in time of the component, and `events` are functions that act on that state (behaviors).
 
 ### Why Models?
 
 Advantages of models:
 * A common API structure to group state and behavior of components
 * Atomic responsibilities
 * Composable and shareable functionality
 
 We use React Hooks to return models. An empty model would look like this:
 
 ```ts
 const useEmptyModel = (config = {}) => {
   const state = {}
   const events = {}
   
   return { state, events }
 }
 ```

A model has configuration for functionality including configuration of events. An event has two parts:
* Guard

> * Action

Author: Explain more of the of the why of these concepts. Also consider listing actions first - also change "actions" to "callbacks"

# Compound Components

## What is a compound component?

> It is a way of describing a complex component by simultaneously describing sub-components. A compound component is made up of a few different parts.
> 
> * Container component
> * Sub-components
> * Shared model (optional)
> * Behavior hooks (optional)
> 
> A compound component contrasts with a configuration component.

Author: I think we need a better definition here and to describe a configuration component.

```suggestion
Compound components is a pattern in which components are used together in such a way that describes a higher level of abstraction. A desktop computer is composed of many components that can be serviced and configured individually. Similarly, a compound component allows simultaneous configuration through access to each sub-component. Here are the parts of a compound component:

* Container component
* Sub-components
* Shared model (optional)
* Behavior hooks (optional)

A compound component contrasts with a configuration component which instead configures from a single interface. A configuration component might be like choosing a desktop computer based on stats - how much RAM or how fast the CPU should be or based on what you want to use it for. A compound component is more like buying the parts individually and assembling yourself.
```

Configuration component:
```tsx
<Tabs items={[
  { title: 'First Tab', content: 'First Tab Contents' },
  { title: 'Second Tab', content: 'Second Tab Contents' },
]} />
```

Compound component:
```tsx
<Tabs>
  <Tabs.List>
    <Tabs.Item>First Tab</Tabs.Item>
    <Tabs.Item>Second Tab</Tabs.Item>
  </Tabs.List>
  <Tabs.Panel>First Tab Contents</Tabs.Panel>
  <Tabs.Panel>Second Tab Contents</Tabs.Panel>
</Tabs>
```

In this example, `Tabs` is the container component and `Tabs.List` is a sub-component.

> Compound components always have a container component and sub-components. Components that do not contain sub-components are not considered compound components. Some compound components might not contain state or behavior. An example might be an `IconButton` which is a button that contains an icon. It might be a compound component only for styling purposes, but doesn't contain any special state or behaviors:

Author: I don't like the way this reads:

> Compound components always have a container component and sub-components. Components that do not contain sub-components are not considered compound components.

Those 2 sentences make the same statement. This is also covered by a statement earlier:

> [Compound components are] a way of describing a complex component by simultaneously describing sub-components.

Those 2 sentences do not directly relate to the rest of the paragraph, which talks about not having a model or behavior. Perhaps remove those sentences or move them somewhere else.

```suggestion
Some compound components might not contain state or behavior. An example might be an `IconButton` which is a button that contains an icon. It might be a compound component only for styling purposes, but doesn't contain any special state or behaviors:
```

```tsx
<IconButton onClick={onClick}>
  <IconButton.Icon icon={icon} />
  <IconButton.Text>Button Text</IconButton.Text>
</IconButton>
```

## Container Components

> A container component is the entry point to a compound component. A container component could represent a real DOM element, or just be a non-element container. For example, the `Pagination` component has a container component that represents a `role=nav` element. The `Tabs` container component, however, does not contain a semantic element.

Author: ```suggestion
A container component is the entry point to a compound component. A container component could represent a real DOM element, or be a non-element container. For example, the `Pagination` component has a container component that represents a `role=nav` element. The `Tabs` container component, however, does not contain a semantic element.
```

If a compound component contains any state or behavior, it will also provide a shared model to sub-components via [React context](https://reactjs.org/docs/context.html). A container component takes props for either the model or configuration for the model. In the `Tabs` compound component example, it might look like this:

```tsx
const TabsModelContext = React.createContext({})

> const useTabsModelContext = () => React.useContext(TabsModelContext)

Author: ```suggestion
// This will be used in sub-components
const useTabsModelContext = () => React.useContext(TabsModelContext)
```

```tsx
const useEllipsisTooltipModel = (config = {}) => {
  return useTooltipModel({
    ...config,
    shouldOpen({ data }) {

      // data has an `element` property
      // `findOverflowElement` returns the element with an overflow style applied
      const element = findOverflowElement(data.element)
      
      // if the scrollWidth is greater than the clientWidth,
      // then the content must be overflowed
      return element.scrollWidth > element.clientWidth
    }
  })
}
```

Models are meant to be composable. For example, a `TabModel` uses a `MenuModel` (which itself uses `ListModel`) and a `ListModel` for a list of panels. `TabsModel` also keeps track of which tab is currently active. This might look like the following:

```ts
const useTabsModel = (config = {}) => {
  // id is used for ARIA attributes
  const id = useUniqueId(config.id)
  const [activeTab, setActiveTab] = React.useState('')
  const menu = useMenuModel(config)
  const panels = useListModel(config)
  
  const state = {
    ...menu.state,
    id,
    activeTab,

>     panels: panels.items,

Reviewer: ```suggestion
    panels: panels.state.items,
```

  E extends Record<string, (...args: any[]) => void
> {
  state: S
  events: E
}
```
    
The Typescript interface of Actions and Guards looks like this:
```ts
type Guard<EventData, State> = ({ data: EventData, state: State}) => boolean
type Action<EventData, State> = ({ data: EventData, state: State}) => void
```

## Behavior Hooks

### What is a Behavior Hook?

A behavior hook usually applies to a sub-component and describes the events and attributes that are applied to a sub-component's element. A behavior hook takes in the model and developer-defined HTML attributes and return a merged object of events and HTML attributes.

### Why Behavior Hooks?

A behavior hook allows us to more easily reuse functionality between components with similar sub-components. They also provide another layer of composition to compound components. They are usually the accessibility layer of a sub-component and not the HTML element directly.

For example, the `MenuModel` contains model's internal state and events, but doesn't handle external DOM events directly. The behavior hook is the glue between the model and DOM elements. A `useMenu` behavior hook might look like this:

```ts

> const useMenu const useMenu = (

Reviewer 2: ```suggestion
const useMenu = (
```

>   model,

Author: ```suggestion
  { state, events },
```

  elemProps = {}
) => {

  React.useEffect(() => {
    const item = state.items.find(({ id }) => state.currentId === id)
    item.ref.current?.focus()
  }, [state.currentId, state.items])

  return {
    onKeyDown(event) {
      // if onKeyDown was provided, call it first
      elemProps.onKeyDown?.(event)

      if (event.key === 'ArrowLeft') {

>          events.previous()
>       }
>       if (event.key === 'ArrowRight') {
>          events.next()

Reviewer 2: ```suggestion
         model.events.previous()
      }
      if (event.key === 'ArrowRight') {
         model.events.next()
```
(or destructure `model` above)

Author: Good catch

Author: Also state above... Perhaps I'll destructure then.

      }
    },
    ...elemProps,
  }
}
```

## Putting it all together

In the `Tabs` component example, the Menu doesn't actually exist as component, but a model and a behavior hook. The `Tab.List` sub-component uses the `TabModel` and the `useMenu` behavior hook to produce the desired sub-component. It looks something like this:

```tsx
const TabList = ({children, ...elemProps}) => {
  const model = useTabsModelContext()
  const menu = useMenu(
    model,
    elemProps,
  )

  return (
    <div
      role="tablist"
      {...menu}
    >
      {children}
    </div>
  )
};
```

There are 2 ways to configure a model.

> The easiest way is to pass model configuration directly to the container component. In the `Tabs` example, it might look like this:

Reviewer 2: This example could use a bit of expanding. It's not immediately clear what `onActivateTab` has to do with the model configuration. I think it's just saying something like "props for each model field are generally available for simple customization on components without having to pass in the entire model", etc.

Author: I agree. There is a line above that says:
>  A model has configuration for functionality including configuration of events.

But it isn't obvious this is the connection. It should be reiterated.

Author: It uses the example above:

https://github.com/Workday/canvas-kit/pull/937/files#diff-a467e1a7dbcaa0ea8af669d262eaab2b94cb1302e12f9d81aaa4117a08872536R185-R214

Author: What about the following?

```suggestion
The easiest way is to pass model configuration directly to the container component. Continuing from the `useTabsModel` example above which has a `onActivateTab` action available on the model configuration, it might look like this:
```

```tsx
<Tabs
  onActivateTab={({data}) => fetch('/api/activatetab' + data.id) }
>
  ...
</Tabs>
```

We expect most use-cases will be covered by the above example, but If you need direct access to the model in your component, you'll have to create the configured model yourself and pass it on to the container component. The allows direct access to state and events of the model. In the `Tabs` example, it might look like this:

```tsx
const MyTabs = () => {

>   const model = useTabModel({

Reviewer: ```suggestion
  const model = useTabsModel({
```

</Tabs>
```

In this example, `Tabs` is the container component and `Tabs.List` is a sub-component.

Some compound components might not contain state or behavior. An example might be an `IconButton` which is a button that contains an icon. It might be a compound component only for styling purposes, but doesn't contain any special state or behaviors:

```tsx
<IconButton onClick={onClick}>
  <IconButton.Icon icon={icon} />
  <IconButton.Text>Button Text</IconButton.Text>
</IconButton>
```

## Container Components

A container component is the entry point to a compound component. A container component could represent a real DOM element, or be a non-element container. For example, the `Pagination` component has a container component that represents a `role=nav` element. The `Tabs` container component, however, does not contain a semantic element.

If a compound component contains any state or behavior, it will also provide a shared model to sub-components via [React context](https://reactjs.org/docs/context.html). A container component takes props for either the model or configuration for the model. In the `Tabs` compound component example, it might look like this:

```tsx

> const TabsModelContext = React.createContext({})

Author: In building a new component, `useTabsModel` and `useTabsModelContext` are very close to each other and return the same type. If you use `useTabsModel` (first autocomplete suggestion) instead of the context version, you'll get a new model instance rather than the shared reference and things don't work the way you expect.

The suggestion is instead to export the `TabsModel` as a context reference and don't export a `useTabsModelContext` hook. This also helps reinforce this is explicitly React Context and not a new model.

# Compound Components

## What is a compound component?

> Compound components is a pattern in which components are used together in such a way that describes a higher level of abstraction. A desktop computer is composed of many components that can be serviced and configured individually. Similarly, a compound component allows simultaneous configuration through access to each sub-component. Here are the parts of a compound component:

Reviewer 3: What do you think about this for an alternative to the opening line here?
```
> Compound components is a pattern where higher level components are composed using smaller components, and you retain control over how those smaller pieces are wired together.
```

Reviewer 3: I would even consider formatting it in that way, as a quote, so it stands out and apart from the elaborating paragraph.

Author: I'm not sure we retain control over how the smaller pieces are wired together (at least not from the component API). It allows configuration of both the overall behavior as well as of individual sub-components by providing direct access to semantic elements.

How about:
```suggestion
> Compound components is a pattern where higher level components are composed using smaller components, and you retain access to all the semantic elements of the higher level component.
```

Reviewer 3: Thanks, that's what I was going for.

* [Container component](#container-components)
* [Sub-components](#sub-components)
* [Shared model (optional, advanced)](#models)
* [Behavior hooks (optional, advanced)](#behavior-hooks)

> A compound component contrasts with a configuration component which instead configures from a single interface. A configuration component might be like choosing a desktop computer based on stats - how much RAM or how fast the CPU should be or based on what you want to use it for. A compound component is more like buying the parts individually and assembling yourself.

Reviewer 3: Consider a slight expansion to: "A compound component contrasts with a configuration component which instead configures from a single interface." ...with "... a single interface, like a configuration object or multiple props in React."

Reviewer 3: I like the PC analogy!

Author: Do you mean like this?

```suggestion
A compound component contrasts with a configuration component which instead configures from a single interface, like a configuration object or multiple props in React. A configuration component might be like choosing a desktop computer based on stats - how much RAM or how fast the CPU should be or based on what you want to use it for. A compound component is more like buying the parts individually and assembling yourself.
```
